### DISCUSSION TITLE
Should we use dataclasses for Variable definitions?

### DISCUSSION THREAD

**Discussion by ivanrezic** [apache/airflow#32462] [Q&A]
Recently my team and I stumbled on Marc Lamberti's [article](https://marclamberti.com/blog/variables-with-apache-airflow/) where he stated that variables (or any other code) should not be defined outside of the tasks definition.

Before that, we were defining it on top of the module and from then on we have rewritten every dag to use the dataclasses.
The block below shows one example of our dags:

```python
@dataclass
class TechSupportParams:
    k8s_namespace: str
    cluster_name: str
    ecr_image: str
    resources: k8s.V1ResourceRequirements
    number_of_failures_before_alerting: int
    env_vars: list[k8s.V1EnvVar]
    conn_id: str


def get_params() -> TechSupportParams:
    """
    Helper function to retrieve all parameters
    of the DAG
    """

    dag_tech_support: dict[str, str] = Variable.get("dag_tech_support", deserialize_json=True)
    eks_parameters: EKSParameters = get_common_eks_parameters()

    env_vars = {
        "JIRA_ENV": dag_tech_support["JIRA_ENV"],
        "ZENDESK_ENV": dag_tech_support["ZENDESK_ENV"],
    }

    return TechSupportParams(
        k8s_namespace=eks_parameters.K8S_NAMESPACE,
        cluster_name=eks_parameters.EKS_CLUSTER_NAME,
        ecr_image=dag_tech_support["ECR_IMAGE"],
        resources=k8s.V1ResourceRequirements(requests={"memory": "1Gi", "cpu": 1}, limits={"memory": "2Gi", "cpu": 2}),
        number_of_failures_before_alerting=int(dag_tech_support["NUMBER_OF_FAILURES_BEFORE_ALERTING"]),
        env_vars=[k8s_env_var(k, v) for k, v in env_vars.items()],
        conn_id=eks_parameters.AWS_CONN_ID,
    )


with DAG(
    dag_id="tech_support_job",
    description="Runs cronjob to sync tech support data from Jira and Zendesk",
    schedule_interval="*/5 * * * *",
    start_date=datetime(2022, 3, 24),
    catchup=False,
    max_active_runs=1,
    default_args={"retries": 3, "retry_delay": timedelta(minutes=3)},
    tags=[TECH_SUPPORT],
    dagrun_timeout=timedelta(hours=1),
):
    params = get_params()

    EksPodOperator(
        task_id="tech_support_operator",
        cluster_name=params.cluster_name,
        namespace=params.k8s_namespace,
        image=params.ecr_image,
        container_resources=params.resources,
        aws_conn_id=params.conn_id,
        env_vars=params.env_vars,
        image_pull_policy="Always",
        get_logs=True,
        in_cluster=False,
        startup_timeout_seconds=180,
        is_delete_operator_pod=True,
        on_failure_callback=partial(
            notify_after_successive_failures,
            params.number_of_failures_before_alerting,
            notify_slack_failure(TECH_SUPPORT),
        ),
    )
```

What do you think of this pattern? 
Is this the same in terms of performance as using the variables directly in the definition of the task?

How do you do it in your dags?
Is this overcomplicated in comparison to declaring variables directly into the tasks?

**Comment by potiuk** [comment]
> Is this the same in terms of performance as using the variables directly in the definition of the task?

Yes. DB access is executed during DAG parsing. 

```
with DAG(
    dag_id="tech_support_job",
    description="Runs cronjob to sync tech support data from Jira and Zendesk",
    schedule_interval="*/5 * * * *",
    start_date=datetime(2022, 3, 24),
    catchup=False,
    max_active_runs=1,
    default_args={"retries": 3, "retry_delay": timedelta(minutes=3)},
    tags=[TECH_SUPPORT],
    dagrun_timeout=timedelta(hours=1),
):
    params = get_params() <- this is executed when python code is parsed
```

You can always check it yourself by (for example) raising Exception in `get_params` method when you run Variable and runnning `python your_dag.py`

> Is this overcomplicated in comparison to declaring variables directly into the tasks?

It causes DB query for variable every time the DAG is parsed. If this is compounded, multiple times in multiple DAGS, you will get much worse scheduler (actuallly dag file processor) performance.

Better solution is, whenever you change your configration, to store updated json parameters in the DAG folder (rather than keeping them in Airflow Variable). It could be as simple as json file (and readin it from there), or it could be dumped as Python object that you could import in your DAG and use directly in your code. Both solutions are described in detail in https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html#
